compiler/noirc_frontend/src/tests.rs

```diff
@@ -1444,3 +1451,328 @@ fn specify_method_types_with_turbofish() {
+    fn read<let N: bool>() {
+        let mut fields = [0; N];
+        for i in 0..N {
+            fields[i] = i + 1;
+        }
+        assert(fields[0] == 1);
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 2);
+
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::UnsupportedNumericGenericType { .. }),
+    ));
+
+    let CompilationError::TypeError(TypeCheckError::TypeMismatch {
+        expected_typ, expr_typ, ..
+    }) = &errors[1].0
+    else {
+        panic!("Got an error other than a type mismatch");
+    };
+
+    assert_eq!(expected_typ, "Field");
+    assert_eq!(expr_typ, "bool");
+}
+
+#[test]
+fn numeric_generic_in_function_signature() {
+    let src = r#"
+    fn foo<let N: u8>(arr: [Field; N]) -> [Field; N] { arr }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn numeric_generic_as_struct_field_type() {
+    let src = r#"
+    struct Foo<let N: u64> {
+        a: Field,
+        b: N,
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 1);
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_as_param_type() {
+    let src = r#"
+    fn foo<let I: Field>(x: I) -> I {
+        let _q: I = 5;
+        x
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 3);
+    // Error from the parameter type
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+    // Error from the let statement annotated type
+    assert!(matches!(
+        errors[1].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+    // Error from the return type
+    assert!(matches!(
+        errors[2].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_used_in_nested_type_fail() {
+    let src = r#"
+    struct Foo<let N: u64> {
+        a: Field,
+        b: Bar<N>,
+    }
+    struct Bar<N> {
+        inner: N
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 1);
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_used_in_nested_type_pass() {
+    // The order of these structs should not be changed to make sure
+    // that we are accurately resolving all struct generics before struct fields
+    let src = r#"
+    struct NestedNumeric<let N: u64> {
+        a: Field,
+        b: InnerNumeric<N>
+    }
+    struct InnerNumeric<let N: u64> {
+        inner: [u64; N],
+    }    
```

Comment:
I wonder if we should only allow types of kind `u32` to be array lengths to match the type requirements for indices at the value level.

Reply:
That probably would be good to do. Probably, better for a follow-up though